customer.cpp

```diff
@@ -1,28 +1,66 @@
 #include "customer.h"
 #include "shop.h"
+#include <iostream>
+#include <chrono>
 
 Customer::Customer(std::string name, Shop* shop)
-    :mCustomerName(name)
+    :futureObj(exitSignal.get_future())
+    ,mCustomerName(name)
     ,pBelongsToShop(shop)
+    ,mTerminated(false)
 {
+    mBarberNotifier = shop->getBarberSemaphore();
+    mCustomersNotifier = shop->getCustomersSemaphore();
     mCustomerThread = std::thread(operating, this);
-    mCustomerThread.join();
 }
 
 void Customer::operating()
 {
-    getHaircut();
-    std::this_thread::sleep_for(std::chrono::seconds(1));
+    while (!stopRequested())
```

Comment:
The client has a linear lifetime. No need for loop and stop request. I think he goes in, if he finds a chair waits, gets a haircut and vanishes. If he doesn't find a chair, he just vanishes. Why the loop?

Reply:
This is logical. My thought was that it is possible that some customers could be waiting (threads active but not served yet) when barber gets a request to stop. In this case one option could be to force the remaining customers to terminate. Another option could be to serve them, do not allow anymore and then the barber can be terminated. Do you agree with this second concept ?

Reply:
I agree. I think serving all the clients sitting in the shop before shutting down makes total sense.